lib/internal/main/test_runner.js

```diff
@@ -0,0 +1,124 @@
+'use strict';
+const { ArrayFrom, Promise, SafeSet } = primordials;
+const {
+  prepareMainThreadExecution
+} = require('internal/bootstrap/pre_execution');
+const { spawn } = require('child_process');
+const { readdirSync, statSync } = require('fs');
+const console = require('internal/console/global');
+const {
+  codes: {
+    ERR_TEST_FAILURE,
+  },
+} = require('internal/errors');
+const test = require('internal/test_runner/harness');
+const { kSubtestsFailed } = require('internal/test_runner/test');
+const { doesPathMatchFilter } = require('internal/test_runner/utils');
+const { join, resolve } = require('path');
+const kFilterArgs = ['--test'];
+
+prepareMainThreadExecution(false);
+markBootstrapComplete();
+
+// TODO(cjihrig): Replace this with recursive readdir once it lands.
+function processPath(path, testFiles, userSupplied = false) {
+  const stats = statSync(path);
+
+  if (stats.isFile()) {
+    if (userSupplied || doesPathMatchFilter(path)) {
+      testFiles.add(path);
+    }
+  } else if (stats.isDirectory()) {
+    if (!userSupplied && path.endsWith('node_modules')) {
+      return;
+    }
+
+    const entries = readdirSync(path);
```

Comment:
Should we worry about very large dir that could overflow the memory?

Reply:
It's a concern, but not something I want to worry about here for a few reasons:

1. There is a recursive readdir() implementation PR in the works. I left a TODO to update to that once it's available.
2. It's a fairly unlikely scenario - there was no streaming readdir() implementation in core for a long time and people were mostly fine.
3. It's an experimental API and not one that handles production traffic.